File: backend/src/middleware/rateLimiter.ts
/**
 * Redis-backed rate limiter
 * Uses INCR + EXPIRE pattern — state is shared across all instances
 */

import { Request, Response, NextFunction } from 'express';
import { redisClient } from '../config/redis';

export interface RateLimiterOptions {
  windowMs?: number;
  maxRequests?: number;
  keyGenerator?: (req: Request) => string;
}

const DEFAULT_WINDOW_MS = 60_000;
const DEFAULT_MAX_REQUESTS = 100;

export function createRateLimiter(options: RateLimiterOptions = {}) {
  const windowMs = options.windowMs ?? DEFAULT_WINDOW_MS;
  const windowSec = Math.ceil(windowMs / 1000);
  const maxRequests = options.maxRequests ?? DEFAULT_MAX_REQUESTS;
  const keyGenerator =
    options.keyGenerator ??
    ((req: Request) => req.ip || req.socket.remoteAddress || 'unknown');

  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    // Fall back gracefully if Redis is not connected
    if (!redisClient.isOpen) {
      next();
      return;
    }

    const key = `rate:${keyGenerator(req)}`;

    try {
      const count = await redisClient.incr(key);

      // Set expiry only on the first request in the window
      if (count === 1) {
        await redisClient.expire(key, windowSec);
      }

      const ttl = await redisClient.ttl(key);
      const resetAt = Math.floor(Date.now() / 1000) + (ttl > 0 ? ttl : windowSec);
      const remaining = Math.max(0, maxRequests - count);

      res.setHeader('X-RateLimit-Limit', String(maxRequests));
      res.setHeader('X-RateLimit-Remaining', String(remaining));
      res.setHeader('X-RateLimit-Reset', String(resetAt));

      if (count > maxRequests) {
        res.status(429).json({
          success: false,
          error: {
            code: 'RATE_LIMIT_EXCEEDED',
            message: 'Too many requests. Please try again later.',
          },
        });
        return;
      }
    } catch {
      // Redis error — fail open (don't block requests)
      next();
      return;
    }

    next();
  };
}

export const defaultRateLimiter = createRateLimiter();

File: backend-go/internal/middleware/ratelimiter.go
package middleware

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/ssco/backend/internal/cache"
)

// RateLimiter returns a Gin middleware backed by Redis INCR+EXPIRE.
// Falls back gracefully (fail-open) if Redis is unavailable.
func RateLimiter(cs *cache.Service, windowMs, maxRequests int) gin.HandlerFunc {
	window := time.Duration(windowMs) * time.Millisecond

	return func(c *gin.Context) {
		key := cache.KeyRateLimit(c.ClientIP())
		ctx := c.Request.Context()

		count, err := cs.Incr(ctx, key)
		if err != nil {
			// Redis unavailable — fail open
			c.Next()
			return
		}

		if count == 1 {
			_ = cs.Expire(ctx, key, window)
		}

		ttl, _ := cs.TTL(ctx, key)
		resetAt := time.Now().Add(ttl).Unix()
		remaining := maxRequests - int(count)
		if remaining < 0 {
			remaining = 0
		}

		c.Header("X-RateLimit-Limit", strconv.Itoa(maxRequests))
		c.Header("X-RateLimit-Remaining", strconv.Itoa(remaining))
		c.Header("X-RateLimit-Reset", strconv.FormatInt(resetAt, 10))

		if int(count) > maxRequests {
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"success": false,
				"error": gin.H{
					"code":    "RATE_LIMIT_EXCEEDED",
					"message": "Too many requests. Please try again later.",
				},
			})
			return
		}

		c.Next()
	}
}
